File: benchmarks/python/string.py
ITERS = 100000

test_string = "The quick brown fox jumps over the lazy dog. This is a test string for benchmarking purposes."
chain_string = "   The quick brown fox jumps over the lazy dog. This is a test string for benchmarking purposes.   "

def bench_count(s, needle, iters):
    total = 0
    for _ in range(iters):
        total += s.count(needle)
    return total

def bench_contains(s, needle, iters):
    total = 0
    for _ in range(iters):
        if needle in s:
            total += 1
    return total

def bench_replace(s, old, new, iters):
    total = 0
    for _ in range(iters):
        result = s.replace(old, new)
        total += len(result)
    return total

def bench_substring(s, iters):
    total = 0
    for i in range(iters):
        start = i % 16
        result = s[start:start + 24]
        total += len(result)
    return total

def bench_split(s, delim, iters):
    total = 0
    for _ in range(iters):
        parts = s.split(delim)
        total += len(parts)
    return total

def bench_upper(s, iters):
    total = 0
    for _ in range(iters):
        result = s.upper()
        total += len(result)
    return total

def bench_lower(s, iters):
    total = 0
    for _ in range(iters):
        result = s.lower()
        total += len(result)
    return total

def bench_reverse(s, iters):
    total = 0
    for _ in range(iters):
        rev = s[::-1]
        total += len(rev)
    return total

def bench_trim(iters):
    padded = "   hello world   "
    total = 0
    for _ in range(iters):
        trimmed = padded.strip()
        total += len(trimmed)
    return total

def bench_chain_window_replace(s, iters):
    total = 0
    for i in range(iters):
        start = i % 16
        result = s.strip()[start:start + 40].replace(" ", "_")
        total += len(result)
    return total

def bench_chain_case_replace(s, iters):
    total = 0
    for _ in range(iters):
        result = s.lower().replace("the", "a").upper()
        total += len(result)
    return total

def bench_chain_trim_reverse(iters):
    padded = "   hello world   "
    total = 0
    for _ in range(iters):
        result = padded.strip()[::-1].replace("l", "L")
        total += len(result)
    return total

if __name__ == "__main__":
    print(f"count checksum: {bench_count(test_string, 'e', ITERS)}")
    print(f"contains checksum: {bench_contains(test_string, 'fox', ITERS)}")
    print(f"replace_same checksum: {bench_replace(test_string, 'the', 'THE', ITERS)}")
    print(f"replace_grow checksum: {bench_replace(test_string, 'dog', 'catapult', ITERS)}")
    print(f"replace_shrink checksum: {bench_replace(test_string, 'benchmarking', 'bench', ITERS)}")
    print(f"substring checksum: {bench_substring(test_string, ITERS)}")
    print(f"split checksum: {bench_split(test_string, ' ', ITERS)}")
    print(f"upper checksum: {bench_upper(test_string, ITERS)}")
    print(f"lower checksum: {bench_lower(test_string, ITERS)}")
    print(f"reverse checksum: {bench_reverse(test_string, ITERS)}")
    print(f"trim checksum: {bench_trim(ITERS)}")
    print(f"chain_window_replace checksum: {bench_chain_window_replace(chain_string, ITERS)}")
    print(f"chain_case_replace checksum: {bench_chain_case_replace(test_string, ITERS)}")
    print(f"chain_trim_reverse checksum: {bench_chain_trim_reverse(ITERS)}")

File: benchmarks/go/string.go
package main

import (
	"fmt"
	"strings"
)

const ITERS = 100000

var testString = "The quick brown fox jumps over the lazy dog. This is a test string for benchmarking purposes."
var chainString = "   The quick brown fox jumps over the lazy dog. This is a test string for benchmarking purposes.   "

func benchCount(s, needle string, iters int) int {
	total := 0
	for i := 0; i < iters; i++ {
		total += strings.Count(s, needle)
	}
	return total
}

func benchContains(s, needle string, iters int) int {
	total := 0
	for i := 0; i < iters; i++ {
		if strings.Contains(s, needle) {
			total++
		}
	}
	return total
}

func benchReplace(s, old, newValue string, iters int) int {
	total := 0
	for i := 0; i < iters; i++ {
		result := strings.ReplaceAll(s, old, newValue)
		total += len(result)
	}
	return total
}

func benchSubstring(s string, iters int) int {
	total := 0
	for i := 0; i < iters; i++ {
		start := i % 16
		result := strings.Clone(s[start : start+24])
		total += len(result)
	}
	return total
}

func benchSplit(s, delim string, iters int) int {
	total := 0
	for i := 0; i < iters; i++ {
		parts := strings.Split(s, delim)
		total += len(parts)
	}
	return total
}

func benchToUpper(s string, iters int) int {
	total := 0
	for i := 0; i < iters; i++ {
		result := strings.ToUpper(s)
		total += len(result)
	}
	return total
}

func benchToLower(s string, iters int) int {
	total := 0
	for i := 0; i < iters; i++ {
		result := strings.ToLower(s)
		total += len(result)
	}
	return total
}

func benchReverse(s string, iters int) int {
	total := 0
	runes := []rune(s)
	for i := 0; i < iters; i++ {
		total += len(reverseRunes(runes))
	}
	return total
}

func reverseRunes(runes []rune) string {
	n := len(runes)
	rev := make([]rune, n)
	for j := 0; j < n; j++ {
		rev[j] = runes[n-1-j]
	}
	return string(rev)
}

func benchTrim(iters int) int {
	padded := "   hello world   "
	total := 0
	for i := 0; i < iters; i++ {
		trimmed := strings.Clone(strings.TrimSpace(padded))
		total += len(trimmed)
	}
	return total
}

func benchChainWindowReplace(s string, iters int) int {
	total := 0
	for i := 0; i < iters; i++ {
		start := i % 16
		trimmed := strings.TrimSpace(s)
		window := strings.Clone(trimmed[start : start+40])
		replaced := strings.ReplaceAll(window, " ", "_")
		total += len(replaced)
	}
	return total
}

func benchChainCaseReplace(s string, iters int) int {
	total := 0
	for i := 0; i < iters; i++ {
		result := strings.ToUpper(strings.ReplaceAll(strings.ToLower(s), "the", "a"))
		total += len(result)
	}
	return total
}

func benchChainTrimReverse(iters int) int {
	padded := "   hello world   "
	total := 0
	for i := 0; i < iters; i++ {
		trimmed := strings.Clone(strings.TrimSpace(padded))
		reversed := reverseRunes([]rune(trimmed))
		result := strings.ReplaceAll(reversed, "l", "L")
		total += len(result)
	}
	return total
}

func main() {
	fmt.Printf("count checksum: %d\n", benchCount(testString, "e", ITERS))
	fmt.Printf("contains checksum: %d\n", benchContains(testString, "fox", ITERS))
	fmt.Printf("replace_same checksum: %d\n", benchReplace(testString, "the", "THE", ITERS))
	fmt.Printf("replace_grow checksum: %d\n", benchReplace(testString, "dog", "catapult", ITERS))
	fmt.Printf("replace_shrink checksum: %d\n", benchReplace(testString, "benchmarking", "bench", ITERS))
	fmt.Printf("substring checksum: %d\n", benchSubstring(testString, ITERS))
	fmt.Printf("split checksum: %d\n", benchSplit(testString, " ", ITERS))
	fmt.Printf("upper checksum: %d\n", benchToUpper(testString, ITERS))
	fmt.Printf("lower checksum: %d\n", benchToLower(testString, ITERS))
	fmt.Printf("reverse checksum: %d\n", benchReverse(testString, ITERS))
	fmt.Printf("trim checksum: %d\n", benchTrim(ITERS))
	fmt.Printf("chain_window_replace checksum: %d\n", benchChainWindowReplace(chainString, ITERS))
	fmt.Printf("chain_case_replace checksum: %d\n", benchChainCaseReplace(testString, ITERS))
	fmt.Printf("chain_trim_reverse checksum: %d\n", benchChainTrimReverse(ITERS))
}
